future/atomicTransactionComposer.go

```diff
@@ -212,56 +221,104 @@ func (atc *AtomicTransactionComposer) AddMethodCall(params AddMethodCallParams)
+		if len(params.ApprovalProgram) == 0 || len(params.ClearProgram) == 0 {
+			return fmt.Errorf("ApprovalProgram and ClearProgram must be provided for an application creation call")
+		}
+	} else if params.OnComplete == types.UpdateApplicationOC {
+		if len(params.ApprovalProgram) == 0 || len(params.ClearProgram) == 0 {
+			return fmt.Errorf("ApprovalProgram and ClearProgram must be provided for an application update call")
+		}
+		if (params.GlobalSchema != types.StateSchema{}) || (params.LocalSchema != types.StateSchema{}) {
+			return fmt.Errorf("GlobalSchema and LocalSchema must not be provided for an application update call")
+		}
+	} else if len(params.ApprovalProgram) != 0 || len(params.ClearProgram) != 0 || (params.GlobalSchema != types.StateSchema{}) || (params.LocalSchema != types.StateSchema{}) {
+		return fmt.Errorf("ApprovalProgram, ClearProgram, GlobalSchema, and LocalSchema must not be provided for a non-creation call")
 	}
 
-	selectorValue := params.Method.GetSelector()
-	encodedAbiArgs := [][]byte{selectorValue}
 	var txsToAdd []TransactionWithSigner
-	var abiArgs []interface{}
-	var abiTypes []abi.Type
+	var basicArgValues []interface{}
+	var basicArgTypes []abi.Type
+	var refArgValues []interface{}
+	var refArgTypes []string
+	refArgIndexToBasicArgIndex := make(map[int]int)
 	for i, arg := range params.Method.Args {
-		if _, ok := abi.TransactionArgTypes[arg.AbiType]; ok {
-			txnAndSigner, ok := params.MethodArgs[i].(TransactionWithSigner)
+		argValue := params.MethodArgs[i]
+
+		if arg.IsTransactionArg() {
+			txnAndSigner, ok := argValue.(TransactionWithSigner)
 			if !ok {
 				return fmt.Errorf("invalid arg type, expected transaction")
 			}
 
-			err := atc.validateTransaction(txnAndSigner.Txn)
+			err := atc.validateTransaction(txnAndSigner.Txn, arg.Type)
 			if err != nil {
 				return err
 			}
 			txsToAdd = append(txsToAdd, txnAndSigner)
 		} else {
-			abiType, err := abi.TypeOf(arg.AbiType)
+			var abiType abi.Type
+			var err error
+
+			if arg.IsReferenceArg() {
+				refArgIndexToBasicArgIndex[len(refArgTypes)] = len(basicArgTypes)
+				refArgValues = append(refArgValues, argValue)
+				refArgTypes = append(refArgTypes, arg.Type)
+
+				// treat the reference as a uint8 for encoding purposes
+				abiType, err = abi.TypeOf("uint8")
+			} else {
+				abiType, err = arg.GetTypeObject()
+			}
 			if err != nil {
 				return err
 			}
 
-			abiArgs = append(abiArgs, params.MethodArgs[i])
-			abiTypes = append(abiTypes, abiType)
+			basicArgValues = append(basicArgValues, argValue)
+			basicArgTypes = append(basicArgTypes, abiType)
 		}
 	}
 
-	// Up to 16 args can be passed to app call. First is reserved for method selector and last is
-	// reserved for remaining args packaged into one tuple.
-	if len(abiArgs) > 14 {
-		tupleTypes := abiTypes[14:]
-		tupleValues := abiArgs[14:]
-		tupleType, err := abi.MakeTupleType(tupleTypes)
+	var foreignAccounts []string
+	var foreignApps []uint64
+	var foreignAssets []uint64
```

Comment:
same here: do we need to constrain on the number of foreign objects (as discussed in JS-sdk)